hw/riscv/boot.c

```diff
@@ -46,15 +46,25 @@ static uint64_t kernel_translate(void *opaque, uint64_t addr)
     }
 }
 
-hwaddr riscv_load_firmware(const char *filename)
+hwaddr riscv_load_firmware(const char *filename, hwaddr ram_start,
+                           uint64_t ram_size)
 {
     uint64_t firmware_entry, firmware_start, firmware_end;
+    int64_t size;
 
     if (load_elf(filename, NULL, NULL,
                  &firmware_entry, &firmware_start, &firmware_end,
                  0, EM_RISCV, 1, 0) < 0) {
-        error_report("riscv_boot: could not load firmware '%s'", filename);
-        exit(1);
+        firmware_entry = ram_start;
+        firmware_start = ram_start;
+
+        size = load_image_targphys(filename, firmware_start, ram_size);
```

Comment:
As far as I can tell, the size on load_image_targphys is the maximum firmware size so we could perhaps add a constant for maximum firmware size. This simplifies the patch as the we no longer have to make assumptions about memory type, however it would add a maximum limit on the firmware size. I think the latter is reasonable. There is a firmware size limit on the HiFive Unleashed iirc. Perhaps some suitable value base on the size of a reasonable kernel and ramdisk.

Reply:
QEMU ROM routines won't overwrite past the end of the memory regions so it doesn't matter too much what the limit is.

Reply:
That makes sense. Since we won't have a problem with overwriting past memory regions, how about 2 GB to fill the complete address range of RV32I systems?